### PR TITLE
Support CodeClass2.Parts returning parts in source generated files 

### PR DESCRIPTION
**Pull request by jasonmalinowski**
Some designers look through the other partial parts of a class to look for methods that match event handler signatures. Here I'm adding limited support so you can get a FileCodeModel and CodeElement for things residing in source generated files.

Unlike regular FileCodeModels that have certain lifetime guarantees and use our COM weak handle magic, this is keeping the implementation simple and just returning a new FileCodeModel any time we need one, under the hopes that the uses of this are rare. If that becomes a problem, we can refactor the CodeModel cache to deal with these, but there is some complexity around when we precisely know the document has gone away. Since CodeModel is very much a legacy API I don't want to spend time adding complexity until we know we need it.

This also doesn't implement any support for raising change events if a generator output changes. Again, if that is needed we can update the diffing code accordingly.

### CODE REVIEW

File: src/VisualStudio/CSharp/Test/CodeModel/AbstractFileCodeElementTests.cs [hunk 1]
```diff
@@ -43,11 +43,6 @@ protected TestWorkspace GetWorkspace()
             return WorkspaceAndCodeModel.workspace;
         }
 
-        private VisualStudioWorkspace GetExtraWorkspaceToDisposeButNotUse()
```
Comment (jasonmalinowski) [inline src/VisualStudio/CSharp/Test/CodeModel/AbstractFileCodeElementTests.cs:46]: The VisualStudioWorkspace instance is now created by the MEF container, which will dispose it when we dispose the container itself.

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs [hunk 2]
```diff
@@ -340,13 +340,20 @@ public override EnvDTE.FileCodeModel GetFileCodeModel(DocumentId documentId)
                 throw new ArgumentNullException(nameof(documentId));
             }
 
-            var documentFilePath = GetFilePath(documentId);
-            if (documentFilePath == null)
+            var document = _threadingContext.JoinableTaskFactory.Run(() => CurrentSolution.GetDocumentAsync(documentId, includeSourceGenerated: true).AsTask());
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:343]: :(

is this on UI thread?  does this mean we're now blocking hte UI thread on unbounded source-generator work?

Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:343]: That's not new, in essence: any CodeModel operation is going to be asking for semantics, which will block.

File: src/VisualStudio/Core/Impl/CodeModel/FileCodeModel.cs [hunk 3]
```diff
@@ -388,6 +400,11 @@ internal bool TryGetDocument([NotNullWhen(true)] out Document? document)
             {
                 document = _previousDocument;
             }
+            else if (_isSourceGeneratedOutput)
+            {
+                document = State.ThreadingContext.JoinableTaskFactory.Run(
+                    () => Workspace.CurrentSolution.GetSourceGeneratedDocumentAsync(GetDocumentId(), CancellationToken.None).AsTask());
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/FileCodeModel.cs:406]: same concern here.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModel.cs [hunk 4]
```diff
@@ -110,5 +109,12 @@ public void OnSourceFileRenaming(string filePath, string newFilePath)
 
         EnvDTE.FileCodeModel IProjectCodeModel.GetOrCreateFileCodeModel(string filePath, object parent)
             => this.GetOrCreateFileCodeModel(filePath, parent).Handle;
+
+        public EnvDTE.FileCodeModel CreateFileCodeModel(SourceGeneratedDocument sourceGeneratedDocument)
+        {
+            // Unlike for "regular" documents, we make no effort to cache these between callers or hold them for longer lifetimes with
+            // events.
+            return FileCodeModel.Create(GetCodeModelCache().State, parent: null, sourceGeneratedDocument.Id, isSourceGeneratorOutput: true, new TextManagerAdapter()).Handle;
```
Comment (CyrusNajmabadi) [inline src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModel.cs:117]: do you want to CWT this?

File: src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs (hunk 2 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.cs:343]: Chatting with @CyrusNajmabadi offline, we also realized that in many cases they only way you get here is already having ran generators. i.e. you won't have known there's a DocumentId for the source generated file in the first place without generators having ran.

File: src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModel.cs (hunk 4 above)
Comment (jasonmalinowski) [reply src/VisualStudio/Core/Impl/CodeModel/ProjectCodeModel.cs:117]: I could, I didn't want to do any caching at all just to keep things simple.

**Pull request merged**